File: talleres/taller03/Taller3.java
public class Taller3
{    
    static void towerOfHannoi(int n){
        towerOfHannoi(n, 1, 2, 3);
    }

    private static void towerOfHannoi(int n, int beg, int aux, int end){
        if(n == 1){
            System.out.println(beg + " → " + end);
        }else{
            towerOfHannoi(n - 1, beg, end, aux);
            towerOfHannoi(1, beg, aux, end);
            towerOfHannoi(n - 1, aux, beg, end);
        }
    }

    static void combinations(String s){
        combinations("", s);
    }

    private static void combinations(String base, String s){
        if(s.length() == 0){
            System.out.println(base);
        }else{
            combinations(base + s.charAt(0), s.substring(1));
            combinations(base + s.charAt(0), s.substring(1));
        }
    }

    static void permutations(String s) {
        permutations("", s);
    }

    private static void permutations(String base, String s){
        if(s.length() == 0){
            System.out.println(base);
        }else{
            for(int i = 0; i < s.length(); ++i){
                permutations(base + s.charAt(i), s.substring(0,i) + s.substring(i + 1));
            }
        }        
    }

    static void permutation4(String a){
        permutations4(a, "", a.length(), a.length());
    }

    private static void permutations4(String s, String prefix, int size, int count){
        if(count == 1){
            for(int i = 0; i < size; i++){
                System.out.println(prefix + s.charAt(i));
            }
        }
        else{
            for(int i = 0; i < size; i++){
                permutations4(s, prefix + s.charAt(i), size, count - 1);
            }
        }
    }

    static void floodFill(int[][] matriz, int x, int y, int originalColor, int replacementColor){
        if(matriz[x][y] == originalColor){
            matriz[x][y] = replacementColor;
            if(x != 0)floodFill(matriz, x - 1, y, originalColor, replacementColor);
            if(y != 0)floodFill(matriz, x, y - 1, originalColor, replacementColor);
            if(x < matriz.length - 1)floodFill(matriz, x + 1, y, originalColor, replacementColor);
            if(y < matriz[0].length - 1)floodFill(matriz, x, y + 1, originalColor, replacementColor);
        }
    }

    static void main(String[] args){
        int[][] a = new int[10][10];
        a[0][0] = 2;
        a[1][0] = 2;
        a[0][1] = 2;
        a[1][1] = 2;    
        floodFill(a, 5, 5, 0, 1);
        for(int[] i: a){
            for(int j: i){System.out.print(j + " ");}
            System.out.println();
        }        
    }
}

File: talleres/taller03/Taller3.py
def towerOfHannoiAux(n, beg, aux, end):
    if n == 1:
        print(str(beg) + "→" + str(end))
    else:
        towerOfHannoiAux(n - 1, beg, end, aux)
        towerOfHannoiAux(1    , beg, aux, end)
        towerOfHannoiAux(n - 1, aux, beg, end)

def towerOfHannoi(n):
    towerOfHannoiAux(n, 1, 2, 3)

##towerOfHannoi(5)

def combinations(s):
    combinationsAux("", s)

def combinationsAux(base, s):
    if len(s) == 0:
        print(base)
    else:
        combinationsAux(base + s[0], s[1:])
        combinationsAux(base, s[1:])

##combinations("abc")

def permutations(s):
    permutationsAux("", s)

def permutationsAux(base, s):
    if len(s) == 0:
        print(base)
    else:
        for i in range(len(s)):
            permutationsAux(base + s[i], s[:i] + s[i + 1:])
    
##permutations("abcd")

def permutations4(s):
    permutations4Aux(s, "", len(s), len(s))

def permutations4Aux(s, prefix, long, count):
    if count == 1:
        for i in range(long):
            print(prefix + s[i])
    else:
        for i in range(long):
            permutations4Aux(s, prefix + s[i], long, count - 1)

##permutations4("abc")

def floodFill(m, i, j, originalColor, replacementColor):
    if m[i][j] == originalColor:
        m[i][j] = replacementColor
        if i != 0:
            floodFill(m, i - 1, j, originalColor, replacementColor)
        if j != 0:
            floodFill(m, i, j - 1, originalColor, replacementColor)
        if i < len(m) - 1:
            floodFill(m, i + 1, j, originalColor, replacementColor)
        if j < len(m) - 1:
            floodFill(m, i, j + 1, originalColor, replacementColor)

m = []
for i in range(10):
    m.append([0] * 10)

for row in m:
    print row

floodFill(m, 3, 3, 0, 1)

for row in m:
    print row
